### PR TITLE
[WIP][SPARK-29334][ML][MLLIB] Support for basic vector operators

### PR DESCRIPTION
**Pull request by phpisciuneri**
### What changes were proposed in this pull request?

Basic operator support for `ml.linalg.Vector`, and `mllib.linagl.Vector`:

- negation
- scalar multiplication and division
- vector addition and subtraction 

### Why are the changes needed?

To improve parity with Vector operators provided with pySpark.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Unit tests in the appropriate places.  To run quickly:

```
sbt "mllib-local/testOnly org.apache.spark.ml.linalg.VectorsSuite"
```

### CODE REVIEW

**Comment by phpisciuneri** [comment]
I've marked this PR as WIP because I am interested in feedback.  There are any number of ways of implementing.  I've wrapped BLAS where I could, and had to write some custom routines for handling `SparseVector` operations.

Additional work on this PR could include:
- [ ] implicit support for a more natural experience.  The scalar operations are one-sided, and appending colon for analogous right associative operators seemed clunky and unintuitive.
- [ ] replicating this code/functionality on the `mllib` side (wanted to wait for feedback before doing that)

**Review by srowen — COMMENTED** [review]
I think this is pretty OK, on the ground of parity with Python methods like `__add__` etc. I think the implementations could be further optimized, but I don't think it's vital now. I wonder are there places in the code that can use these operations for clarity?

File: mllib-local/src/main/scala/org/apache/spark/ml/linalg/Vectors.scala [hunk 1]
```diff
@@ -186,6 +186,46 @@ sealed trait Vector extends Serializable {
    */
   @Since("3.0.0")
   def dot(v: Vector): Double = BLAS.dot(this, v)
+
+  /**
```
Comment (zhengruifeng) [inline mllib-local/src/main/scala/org/apache/spark/ml/linalg/Vectors.scala]: what about also adding `def +(a: Double)`, `def -(a: Double)` for scalar input and `def *(v: Vector)`, `def /(v: Vector)` as element-wise ops?

Comment (srowen) [reply mllib-local/src/main/scala/org/apache/spark/ml/linalg/Vectors.scala]: Seems reasonable. I'd generally prefer to focus on parity with Python; is that what __add__ supports for example?

Comment (phpisciuneri) [reply mllib-local/src/main/scala/org/apache/spark/ml/linalg/Vectors.scala]: I am not a python expert by any means, but I think this type of behavior is automatically supported with [broadcasting](https://docs.scipy.org/doc/numpy/reference/generated/numpy.add.html)?  @zhengruifeng Can you verify?

If so I am happy to implement.

**Comment by phpisciuneri** [comment]
@srowen @zhengruifeng Thanks for the feedback!

> I think the implementations could be further optimized, but I don't think it's vital now

Yep, I definitely agree.  When the time comes we can try a few different implementations and do a scaling analysis.

> I wonder are there places in the code that can use these operations for clarity?

I will search for usage of the wrapped BLAS routines and see.  One thing to note, for example, is that `BLAS.axpy` is a function with side effects (the `y` arg is modified).  So depending on how the routine is used in code, it might not make sense to substitute unless the `y` arg is copied before the call.

File: mllib-local/src/main/scala/org/apache/spark/ml/linalg/Vectors.scala (hunk 1 above)
Comment (zhengruifeng) [reply mllib-local/src/main/scala/org/apache/spark/ml/linalg/Vectors.scala]: Sorry for late reply.
Yes numpy support broadcasting.
But I guess what @srowen means is those `Vector`/`Vectors` class defined in 'spark/python/pyspark/ml/linalg/__init__.py'

File: mllib-local/src/main/scala/org/apache/spark/ml/linalg/Vectors.scala [hunk 2]
```diff
@@ -753,6 +799,24 @@ class SparseVector @Since("2.0.0") (
...
   }
+
+  def +(v: Vector): Vector = {
+    require(this.size == v.size)
+    v match {
+      case dv: DenseVector =>
+        val arr = dv.values.clone
+        for (i <- this.indices) arr(i) += this(i)
+        Vectors.dense(arr)
+      case sv: SparseVector =>
```
Comment (zhengruifeng) [inline mllib-local/src/main/scala/org/apache/spark/ml/linalg/Vectors.scala]: Current impl: 1, needs distinct and sort, 2, complexity = O(nnz * log(nnz))
It maybe faster to add two `SparseVector` in this way:
```scala
  def add(v: SparseVector): SparseVector = {
    require(size == v.size)
    if (v.indices.isEmpty) {
      copy
    } else if (indices.isEmpty) {
      v.copy
    } else {
      val thisIndices = indices
      val thatIndices = v.indices
      val thisValues = values
      val thatValues = v.values

      val indicesBuilder = mutable.ArrayBuilder.make[Int]
      val valuesBuilder = mutable.ArrayBuilder.make[Double]
      var thisCur = 0
      var thatCur = 0

      while (thisCur < thisIndices.length && thatCur < thatIndices.length) {
        val thisIndex = thisIndices(thisCur)
        val thatIndex = thatIndices(thatCur)
        val (index, value) = if (thisIndex == thatIndex) {
          val sum = thisValues(thisCur) + thatValues(thatCur)
          thisCur += 1
          thatCur += 1
          (thisIndex, sum)
        } else if (thisIndex < thatIndex) {
          val sum = thisValues(thisCur)
          thisCur += 1
          (thisIndex, sum)
        } else {
          val sum = thatValues(thatCur)
          thatCur += 1
          (thatIndex, sum)
        }

        if (value != 0.0) {
          indicesBuilder += index
          valuesBuilder += value
        }
      }

      if (thisCur < thisIndices.length) {
        while (thisCur < thisIndices.length) {
          val index = thisIndices(thisCur)
          val value = thisValues(thisCur)
          if (value != 0.0) {
            indicesBuilder += index
            valuesBuilder += value
          }
          thisCur += 1
        }
      } else if (thatCur < thatIndices.length) {
        while (thatCur < thatIndices.length) {
          val index = thatIndices(thatCur)
          val value = thatValues(thatCur)
          if (value != 0.0) {
            indicesBuilder += index
            valuesBuilder += value
          }
          thatCur += 1
        }
      }

      new SparseVector(size, indicesBuilder.result(), valuesBuilder.result())
    }
  }
```

Comment (zhengruifeng) [reply mllib-local/src/main/scala/org/apache/spark/ml/linalg/Vectors.scala]: I think we can leave it alone at first.

Comment (zhengruifeng) [reply mllib-local/src/main/scala/org/apache/spark/ml/linalg/Vectors.scala]:
```scala
    val sum = mutable.OpenHashMap.empty[Int, Double] // or org.apache.spark.util.collection.OpenHashMap
    this.foreachActive { case (i, v) => sum.update(i, v) }
    o.foreachActive { case (i, v) => val s = sum.getOrElse(i, 0.0) + v; sum.update(i, s) }
    Vectors.sparse(size, sum.toSeq.filter(_._2 != 0))
```

Comment (zhengruifeng) [reply mllib-local/src/main/scala/org/apache/spark/ml/linalg/Vectors.scala]: using foreachActive should be faster than collecting the indices and call apply for each index.

File: mllib-local/src/main/scala/org/apache/spark/ml/linalg/Vectors.scala [hunk 3]
```diff
@@ -186,6 +186,46 @@ sealed trait Vector extends Serializable {
...
+  }
+
+  /**
+   * Divide a vector by a scalar.
+   */
+  @Since("3.0.0")
+  def /(a: Double): Vector = {
+    a match {
+      case 0 => throw new IllegalArgumentException("division by zero")
+      case _ => this * (1 / a)
```
Comment (zhengruifeng) [inline mllib-local/src/main/scala/org/apache/spark/ml/linalg/Vectors.scala]: `case 1 => this` ?

**Pull request closed without merging**